### DISCUSSION TITLE
OOM with RDKit_minimal

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#4066] [General]
Hi,
I'm not sure if this is a bug or feature, but I get an OOM error when processing a large number of SMILES strings with the code below.  Whilst processing molecule 1435, it gives me

Unhandled Rejection (RuntimeError): abort(OOM). Build with -s ASSERTIONS=1 for more info. (evaluating 'new WebAssembly.RuntimeError(what)')

in the 'let match = ....' line, presumably when doing get_substruct_match.

I'm running in Safari on a MacBook Pro under Catalina, using 2021_03_1.  I compiled the MinimalLib myself using the docker file provided in Code/MinimalLib.  It goes down in a similar place using Firefox on the same machine.

I don't know enough about WASM or Javascript to know if this sort of memory use is expected, or if it's maybe a memory leak.

Cheers,

Dave
 
            let all_smarts = smarts.split(' ');
            const qmols = all_smarts.map(smt => {
                return this.state.rdkit.get_qmol(smt);
            });
            let match_mols = [];
            let no_match_mols = [];
            this.state.all_mols.forEach(mol => {
                console.log('MOL : ' + mol);
                let rmol = this.state.rdkit.get_mol(mol[1]);
                let matched = false;
                qmols.forEach(qmol => {
                    if (!matched) {
                        let match = JSON.parse(rmol.get_substruct_match(qmol));
                        if (match.atoms && match.atoms.length) {
                            matched = true;
                        }
                    }
               });
               if (matched) {
                   match_mols.push(mol);
               } else {
                   no_match_mols.push(mol);
               }
            });

**Comment by DavidACosgrove** [comment]
Some more reading and experiment suggests that this is because, although the underlying RDKit RWMol is being managed by a unique_ptr, so there is no memory leak as such, the garbage collector isn't keeping up so it looks like a memory leak.  I will send in a PR with an explicit delete_mol() function that resets the unique_ptr, for those desperate times when needs must.

**Reply by ptosco** [reply]
@DavidACosgrove Hi David, maybe I am misunderstanding the problem but wouldn't calling `delete()` on the `mol` object on the JS side solve the OOM issue? `delete()` _must_ be called on all C++ objects to avoid leaking memory:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management
Calling `delete()` on the JS side should have the same effect as resetting the `unique_ptr` on the C++ side.

**Reply by DavidACosgrove** [reply]
@ptosco Hi Paolo, I didn't find it defined in the JSMol object.  Does it come for free with Emscripten?  I confess, I didn't try it.

**Reply by DavidACosgrove** [reply]
OK, so I've learned something :-).  I will attempt to withdraw my foolish PR.

**Reply by ptosco** [reply]
@DavidACosgrove Yes, it comes for free with Emscripten with every C++ object. Glad that it solves the problem!

**Comment by DavidACosgrove** [comment]
Thanks for the help.

On Wed, 21 Apr 2021 at 16:45, Paolo Tosco ***@***.***> wrote:

> @DavidACosgrove <https://github.com/DavidACosgrove> Yes, it comes for
> free with Emscripten with every C++ object. Glad that it solves the problem!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4066#discussioncomment-640736>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FRI2KCSPHFEOTW5U6DTJ3XJVANCNFSM43J7NT3A>
> .
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk
